### PR TITLE
Remove iterator from collection in favour of fetchAll

### PR DESCRIPTION
**Pull request by Cypher1**
This is a step towards moving from `arcs.sdk.jvm.CollectionImpl` to `arcs.core.storage.handle.CollectionImpl` and  `arcs.sdk.jvm.SingletonImpl` to `arcs.core.storage.handle.SingletonImpl`.

### CODE REVIEW

File: javatests/arcs/sdk/wasm/CollectionApiTest.kt [hunk 1]
```diff
@@ -70,7 +70,7 @@ class CollectionApiTest : AbstractCollectionApiTest() {
                 handles.outHandle.store(d2)
 
                 // Ranged iteration; order is not guaranteed so use 'num' to assign sorted array slots.
-                val sorted = handles.ioHandle.sortedBy { it.num.toInt() }
+                val sorted = handles.ioHandle.fetchAll().sortedBy { it.num.toInt() }
```
Comment (Cypher1) [inline javatests/arcs/sdk/wasm/CollectionApiTest.kt]: @SHeimlich Maybe we should look into having a `.fetchAllBy` to make this nicer?

File: javatests/arcs/sdk/wasm/HandleSyncUpdateTest.kt [hunk 2]
```diff
@@ -28,7 +28,8 @@ class HandleSyncUpdateTest : AbstractHandleSyncUpdateTest() {
         if (handle.name == "sng") {
             num = handles.sng.fetch()?.num ?: -1.0
         } else if (handle.name == "col") {
-            num = if (handles.col.size > 0) handles.col.iterator().next().num else -1.0
+            // TODO(cypher1): This is unsafe, there is no guarantee of ordering.
+            num = if (handles.col.size > 0) handles.col.fetchAll().iterator().next().num else -1.0
```
Comment (Cypher1) [inline javatests/arcs/sdk/wasm/HandleSyncUpdateTest.kt]: @csilvestrini Is this expected user code? I have added a comment as I don't believe this is stable, but would like to know what is expected.

File: particles/Tutorial/Kotlin/Demo/src/total/TTTGame.kt [hunk 3]
```diff
@@ -141,7 +141,7 @@ class TTTGame : AbstractTTTGame() {
         ))
     }
 
-    private fun hasReset() = handles.events.any { it.type == "reset" }
+    private fun hasReset() = handles.events.fetchAll().any { it.type == "reset" }
```
Comment (Cypher1) [inline particles/Tutorial/Kotlin/Demo/src/total/TTTGame.kt]: This pattern might be a good target for a query in future.

File: javatests/arcs/sdk/wasm/UnicodeTest.kt [hunk 4]
```diff
@@ -19,7 +19,8 @@ class UnicodeTest : AbstractUnicodeTest() {
         val pass = if (handle.name == "sng") {
             ((handle as WasmSingletonImpl<*>).fetch() as UnicodeTest_Sng).pass
         } else {
-            ((handle as WasmCollectionImpl<*>).iterator().next() as UnicodeTest_Col).pass
+            // TODO(cypher1): This is unsafe, there is no guarantee of ordering.
+            ((handle as WasmCollectionImpl<*>).fetchAll().iterator().next() as UnicodeTest_Col).pass
```
Comment (csilvestrini) [inline javatests/arcs/sdk/wasm/UnicodeTest.kt]: I'm not too familiar with this test case, but I suspect we only expect there to be a single element in the collection.

If that's true, I think you can use `.single()` to return the single element or throw: `handle.fetchAll().single()`

**Comment by Cypher1** [comment]
Closed as mentioned by @csilvestrini

**Pull request closed without merging**